packages/vgplot/test/render.test.js

```diff
@@ -0,0 +1,58 @@
+import { beforeEach, afterEach, expect, describe, it } from 'vitest';
+import { readFile } from 'node:fs/promises';
+import { resolve } from 'node:path';
+import { Coordinator } from '@uwdata/mosaic-core';
+import { JSDOM } from 'jsdom';
+import { nodeConnector } from './util/node-connector.js';
+import { createAPIContext } from '../src/index.js';
+import { clientsReady } from './util/clients-ready.js';
+
+const cwd = import.meta.dirname;
+
+beforeEach(() => {
+  const dom = new JSDOM(
+    `<!DOCTYPE html><body></body>`,
+    { pretendToBeVisual: true }
+  );
+
+  globalThis.window = dom.window;
+  globalThis.document = dom.window.document;
+  globalThis.navigator ??= dom.window.navigator;
+  globalThis.requestAnimationFrame = window.requestAnimationFrame;
+});
+
+afterEach(() => {
+  // remove browser environment globals
+  delete globalThis.requestAnimationFrame;
+  if (globalThis.navigator === globalThis.window.navigator) {
+    delete globalThis.navigator;
+  }
+  delete globalThis.document;
+  delete globalThis.window;
+});
+
+describe('render', () => {
+  it('should render the density1d spec', () => {
+    return renderTest('density1d');
+  });
+  it('should render the airline-travelers spec', () => {
+    return renderTest('airline-travelers');
+  });
+  it('should render the weather spec', () => {
+    return renderTest('weather');
+  });
+});
+
+async function renderTest(name) {
+  const specPath = resolve(cwd, `specs/${name}.js`);
+  const htmlPath = resolve(cwd, `output/${name}.html`);
+  const [expected, { default: run }] = await Promise.all([
+    readFile(htmlPath, { encoding: 'utf8' }),
+    import(specPath)
+  ]);
+  const mc = new Coordinator(nodeConnector(), { logger: null });
+  const el = await run(createAPIContext({ coordinator: mc }));
+
+  await clientsReady(el);
+  expect(el.outerHTML).toEqual(expected);
```

Comment:
Might be easier to use `toMatchSnapshot` (https://vitest.dev/guide/snapshot) especially if we want to add more tests like this in the future.

Reply:
Ooh neat. This looks like the right method to use in the future: https://vitest.dev/guide/snapshot#file-snapshots